### PR TITLE
Account Features

### PR DESCRIPTION
**Pull request by sarveshpro**
- [x] Restructure the profile and add menu on the left sidebar
- [x] Disable "Projects" menu for all other account types other than TPO.
- [x] Being able to access "Create Widget" was limited to public profiles before and showed a dialogue asking the user if the profile should be made public before opening the widget creator.
- [x] Some screens as "Edit Profile" and "Register Trees" which had been designed to be displayed centered, do not look good if they are just shown aligned to the right border of the menu.
- [x] "Set target" and "Redeem" from the profile screen maybe later on can also be converted into menu items, but they would of course also do not look good the way they are currently designed as modal dialogues.
- [x] "Register Trees" maybe could be removed to avoid redundancy.
- [x] Token refresh
- [x]  Sidebar features
- [x] Drawer open and close on hover
- [x] Close below 768px
- [x] Close submenu of other options
- [x] Login redirect when we user access private account

### CODE REVIEW

**Comment by norbertschuler** [comment]
- [ ] I don't think it makes sense to offer a certificate download for a failed donation - either the backend should not return a link for it (which by the way should not contain `../../` in the path name) or the frontend should not show the download link for a non successful state (need to have knowledge of all the possible states for this of course):

Bildschirmfoto 2021-06-18 um 16 33 21

**Comment by norbertschuler** [comment]
- [ ] Manage projects currently exists twice - as an own tab in the account section and also under the TPO profile - which one do we want to keep (or both)? Also the pages to add or edit a project to do use the account header. Is this wanted or missing?

File: public/static/locales/en/me.json [hunk 1]
```diff
@@ -60,6 +60,10 @@
   "download": "Download",
   "donations": "Donations",
   "in-progress": "In Progress",
+  "donation-in-progress": "In Progress",
+  "donation-tree-cash": "Tree Cash",
+  "donation-success": "Successful",
+  "donation-cancelled": "Successful",
```
Comment (norbertschuler) [inline public/static/locales/en/me.json:66]: Shouldn't this be `Cancelled`?

File: src/features/user/Account/components/TreeMapper/Map.tsx [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+          ...viewport,
+          center: [longitude, latitude],
+          zoom: [zoom],
+        };
+        setViewPort(newViewport);
+      } else {
+        for (const key in geoJson.features) {
+          if (Object.prototype.hasOwnProperty.call(geoJson.features, key)) {
+            const feature = geoJson.features[key];
+            console.log(feature, selectedLocation);
```
Comment (norbertschuler) [inline src/features/user/Account/components/TreeMapper/Map.tsx:219]: `console.log` should be removed before merging.

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 3]
```diff
@@ -246,203 +246,193 @@ export default function RegisterTrees({}: Props) {
...
+                          <DatePicker
+                            label={t('me:datePlanted')}
+                            value={properties.value}
+                            onChange={properties.onChange}
+                            inputVariant="outlined"
+                            TextFieldComponent={MaterialTextField}
+                            autoOk
+                            disableFuture
+                            minDate={new Date(new Date().setFullYear(1950))}
+                            format="MMMM d, yyyy"
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:321]: This is not the preferred date format for Germany. I am not sure if we have already localized date formats yet in any other part of the our web app code (at least we did it for the native app and the old web client)? If yes, we should adapt it the same way, otherwise introducing language depending date formatting can be introduced in another PR e.g. by using i18n.

**Review by norbertschuler — CHANGES_REQUESTED** [review]
This is already looking quite good. I found some minor issues (also see comments above).

**Comment by sagararyal** [comment]
#### 📝Changed routes:


| [**/account/history**](https://planet-webapp-saxgy8os6-planetapp.vercel.app/account/history) | [**/account/manage-projects**](https://planet-webapp-saxgy8os6-planetapp.vercel.app/account/manage-projects) |
|:-:|:-:|
|<a href="https://planet-webapp-saxgy8os6-planetapp.vercel.app/account/history">Screenshot of /account/history</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/lSRd0RNpPosuVA7LgeGIgNQeIQNy1Hn5iK76vyl7jukhOiiID8p4GaOFbjy1bfnGa-92Zn7wvkv2h19bX8c4Vg.png">(view full size)</a> |<a href="https://planet-webapp-saxgy8os6-planetapp.vercel.app/account/manage-projects">Screenshot of /account/manage-projects</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/vr_mF4vTK92T8cKu_WAUOQ5wnOGrOCbApNwvZsdt1HPAqNQsPCjQyofST7axBzLM099ekVEEZMIP47IO-SLugg.png">(view full size)</a> |



| [**/account/register-trees**](https://planet-webapp-saxgy8os6-planetapp.vercel.app/account/register-trees) | [**/account/treemapper**](https://planet-webapp-saxgy8os6-planetapp.vercel.app/account/treemapper) |
|:-:|:-:|
|<a href="https://planet-webapp-saxgy8os6-planetapp.vercel.app/account/register-trees">Screenshot of /account/register-trees</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/aVwvc3p1TuCcmGoWolQq_ysq3jfxcKtAoKdhsjIps9I6f0wE5HzfdY3Y7QTXjBLb_t3Ep9Rr-CNGLX1QjjSFNA.png">(view full size)</a> |<a href="https://planet-webapp-saxgy8os6-planetapp.vercel.app/account/treemapper">Screenshot of /account/treemapper</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/5i6CfeMCDYapGLafig6c3MB69X_j8l495Cw4NLFzWamO-jDcC8Zfo9cGVkSM4dBi-gue0AtI8_aV1hp1Fgwrjg.png">(view full size)</a> |







Commit 95723aa83951d26a816ce6d4063b6ef8b631e94a (https://planet-webapp-saxgy8os6-planetapp.vercel.app).

**Comment by sagararyal** [comment]
@sarveshpro visiting https://planet-webapp-pznk8iory-planetapp.vercel.app/account/treemapper should redirect the user to login (if not logged in) and when completed, open the page.

**Comment by harshvitra** [comment]
Please do not test with the production URL, it will populate data in the production DB

**Comment by harshvitra** [comment]
Screenshot 2021-06-29 at 10 39 24 AM

The Add project is not in the center and manageProjects label is not showing resource

**Comment by harshvitra** [comment]
Screenshot 2021-06-29 at 10 44 44 AM

Clicking on Manage projects does not work

**Comment by harshvitra** [comment]
If there is no access, it is stuck on the loader

https://planet-webapp-pznk8iory-planetapp.vercel.app/manage-projects/proj_WWbJvQMd4FhlrhJI870a54H5

Screenshot 2021-06-29 at 10 47 07 AM

**Comment by harshvitra** [comment]
Seeing someone else's account shows the loader on the right

Screenshot 2021-06-29 at 10 48 00 AM

**Comment by sagararyal** [comment]
At the current state, 2FA requirement throws an error:
Screen Shot 2021-08-21 at 8 59 38 AM

**Comment by sagararyal** [comment]
#### 📝 Changed routes:


| [**/c/\[id\]**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/c/[id]) | [**/claim/\[type\]/\[code\]**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/claim/[type]/[code]) |
|:-:|:-:|
|<a href="https://planet-webapp-8g0w35h3z-planetapp.vercel.app/c/[id]">Screenshot of /c/[id]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/cVqWXOdmwQUrxnN5xK3M3agkrCdsanMAeJqPdH24MKfLmH87zVBxlx1D4zlIHuf3-mBaahuBpDbaI2C-qmU2CA.png">(view full size)</a> |<a href="https://planet-webapp-8g0w35h3z-planetapp.vercel.app/claim/[type]/[code]">Screenshot of /claim/[type]/[code]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/MBDc93MR2yNfmuThR2smrHMmEo0J6PzWKBIrFU0iX-ZZTLIz5p8KkPeyl5p02Lsn18tKWX1PYfTLSTOpXa_8kA.png">(view full size)</a> |



| [**/manage-projects/\[id\]**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/manage-projects/[id]) | [**…e-projects/add-project**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/manage-projects/add-project) |
|:-:|:-:|
|<a href="https://planet-webapp-8g0w35h3z-planetapp.vercel.app/manage-projects/[id]">Screenshot of /manage-projects/[id]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/pSndLNtSIXvFUlqbeR4qReIrJ5vuawcWpUGUTklheV9Jeo7tuUe9pG_IZGZr57JXTtM-7xZf32CgLeBxYAltgQ.png">(view full size)</a> |<a href="https://planet-webapp-8g0w35h3z-planetapp.vercel.app/manage-projects/add-project">Screenshot of /manage-projects/add-project</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/7xl1pDVlsNok7LOaYtNUfUxq3XTAuW4zmQ3aoMGtvR9oQSfB8aoNr6TrpaNAnoK4voJWiTXQPFqeiedDZQLLcQ.png">(view full size)</a> |



| [**/profile/delete-account**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/delete-account) | [**/profile/edit**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/edit) |
|:-:|:-:|
|<a href="https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/delete-account">Screenshot of /profile/delete-account</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/ItnHds4PRj-ZjF61fliYwzwO205R_PRMqy38aZfQgoSQK9mPhAMnxfSg2-OX49iCZQqpZgD19TPbzvsA1EXvjA.png">(view full size)</a> |<a href="https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/edit">Screenshot of /profile/edit</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/LpgrMdpKo1tfXT50R1Dynibi4QCHYqXuDVPcqktSw-eTRSSFJ26scNBRTeeybDxFXO3pni-Rke80KruGyZAaKg.png">(view full size)</a> |



And 9 other routes:
- [**/profile/forest**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/forest)
- [**/profile/history**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/history)
- [**/profile**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile)
- [**/profile/projects**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/projects)
- [**/profile/register-trees**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/register-trees)
- [**/profile/treemapper**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/treemapper)
- [**/profile/widgets**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/profile/widgets)
- [**/register-trees**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/register-trees)
- [**/t/\[id\]**](https://planet-webapp-8g0w35h3z-planetapp.vercel.app/t/[id])



Commit b7909e6af2a66021a5b7aea1c7677b7a39133206 (https://planet-webapp-8g0w35h3z-planetapp.vercel.app).

**Comment by norbertschuler** [comment]
> * [ ]  "Register Trees" maybe could be removed to avoid redundancy.

- [ ]  Clicking on "Register Trees" on the remaining icon on the profile screen leads to a 404 page due to a 404 error from the API call https://app-staging.plant-for-the-planet.org/app/projects/account?_scope=extended&currency=EUR (without Auth0 token header) on my staging TPO profile.
Bildschirmfoto 2021-08-24 um 09 12 34

**Comment by norbertschuler** [comment]
- [ ] Please fix lint errors from `npm run lint:errors`

**Comment by norbertschuler** [comment]
- [ ] Tree mappings (or Plant locations?) are not yet implemented, right?

Bildschirmfoto 2021-08-24 um 09 52 13

**Comment by norbertschuler** [comment]
- [ ] As I wrote before, the warning dialog for private profiles before showing the widget editor is missing (next to the possibility to change the own profile to public before proceeding to creating a widget).

**Comment by norbertschuler** [comment]
- [ ] We should not show the payment method given from the API like this:

Bildschirmfoto 2021-08-24 um 10 08 58

**Comment by norbertschuler** [comment]
- [ ] I have a test case on the develop backend with status=`paid` - is this legacy data from my testing or can this still be a valid state? The state is displayed as `Successful` is the frontend:

Bildschirmfoto 2021-08-24 um 10 19 24

But when I filter for `Successful` in the frontend, it is not shown anymore as the backend API call does not return it:

Bildschirmfoto 2021-08-24 um 10 22 18

**Comment by norbertschuler** [comment]
- [ ] Check if we want to do #1202 also in this pull request.

**Comment by norbertschuler** [comment]
- [ ] I tested with my own TPO account and got an error looking at my plant locations:

Bildschirmfoto 2021-09-07 um 08 55 20

**Comment by norbertschuler** [comment]
- [ ] Can we make the dropdown menu items be dark in dark mode:

Bildschirmfoto 2021-09-07 um 08 54 35Bildschirmfoto 2021-09-07 um 08 59 45

**Comment by norbertschuler** [comment]
Bildschirmfoto 2021-09-07 um 18 55 21

- [ ] If you cancel this dialog, you are shown a blank page just showing the navigation elements.

Bildschirmfoto 2021-09-07 um 18 59 57

- [ ] If you confirm this dialog you are redirected to the widget builder without showing any navigation elements any more around it.

Bildschirmfoto 2021-09-07 um 19 00 20

**Comment by sagararyal** [comment]
> The confirmation dialog for changing private profiles to public if you want to open the widget builder has two issues:
> 
> Bildschirmfoto 2021-09-07 um 18 55 21
> 
> * [ ]  If you cancel this dialog, you are shown a blank page just showing the navigation elements.
> 
> Bildschirmfoto 2021-09-07 um 18 59 57
> 
> * [ ]  If you confirm this dialog you are redirected to the widget builder without showing any navigation elements any more around it.
> 
> Bildschirmfoto 2021-09-07 um 19 00 20

I suggest removing cancel button there. that way the UI won't have to change for users who do not want to make their profile public.

File: src/features/user/UserProfile/components/RegisterTrees.tsx (hunk 3 above)
Comment (sarveshpro) [reply src/features/user/UserProfile/components/RegisterTrees.tsx]: We have a function `formatDate()` that I believe formats the dates. But here in the date picker I think this is where we define the locale, not sure if it is working - https://github.com/Plant-for-the-Planet-org/planet-webapp/pull/1116/files/#:~:text=%3CMuiPickersUtilsProvider,%3E

Comment (sarveshpro) [reply src/features/user/UserProfile/components/RegisterTrees.tsx]: Let's discuss this in a new issue as this is not exclusive to this PR

**Comment by sarveshpro** [comment]
> * [ ]  Can we make the dropdown menu items be dark in dark mode:
> 
We are disabling dark mode for this release.

**Comment by norbertschuler** [comment]
- [ ] Testing with localhost against the develop backend I have a tree mapping in my TPO account with a project link which opens at http://localhost:3000/proj_huo41YATepXOif9Jqv7M5vFn leading to an 404 in the API as we probably do not support this kind of links for projects (https://app-development.plant-for-the-planet.org/app/projects/proj_huo41YATepXOif9Jqv7M5vFn?_scope=extended&currency=EUR). The project slug should be used for the link. And also for the users showing the real name of the project would be nicer to read.

Bildschirmfoto 2021-09-08 um 13 28 30

**Comment by norbertschuler** [comment]
- [ ] Failed payments for my test account on staging are not shown when filtering for failed payments (https://app-staging.plant-for-the-planet.org/app/paymentHistory?filter=failed&limit=15):

Bildschirmfoto 2021-09-08 um 14 08 08
Bildschirmfoto 2021-09-08 um 14 08 23

**Comment by norbertschuler** [comment]
- [ ] My trees registration using a polygon is not displayed on the map - I assume we do not support this yet?

Bildschirmfoto 2021-09-08 um 16 26 41

File: src/features/user/TreeMapper/components/Map.tsx [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    ],
+    {
+      padding: {
+        top: isMobile ? 50: 100,
+        bottom: isMobile ? 200 : 100,
+        left: isMobile ? 50 : 100,
+        right: isMobile ? 50 : 100,
+      },
+    }
+  );
```
Comment (norbertschuler) [inline src/features/user/TreeMapper/components/Map.tsx:132]: This code has the risk of crashes in `WebMercatorViewport` as we have tried to fix them at https://github.com/Plant-for-the-Planet-org/planet-webapp/blob/fb9a7002e3950045d5db57794440df4d27a5b1a0/src/utils/maps/zoomToProjectSite.ts#L13-L25

File: src/features/user/TreeMapper/components/TreeMapperList.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+          <TransactionListLoader />
+          <TransactionListLoader />
+          <TransactionListLoader />
+          <TransactionListLoader />
+          <TransactionListLoader />
+          <TransactionListLoader />
+          <TransactionListLoader />
+          <TransactionListLoader />
+          <TransactionListLoader />
+          <TransactionListLoader />
```
Comment (norbertschuler) [inline src/features/user/TreeMapper/components/TreeMapperList.tsx:53]: This looks funny :-)

File: src/features/user/TreeMapper/components/Map.tsx (hunk 4 above)
Comment (norbertschuler) [reply src/features/user/TreeMapper/components/Map.tsx:132]: I also have not added these checks at https://github.com/Plant-for-the-Planet-org/planet-webapp/blob/fb9a7002e3950045d5db57794440df4d27a5b1a0/src/utils/maps/plantLocations.ts#L14

**Review by norbertschuler — CHANGES_REQUESTED** [review]

File: src/features/user/TreeMapper/components/TreeMapperList.tsx (hunk 5 above)
Comment (sarveshpro) [reply src/features/user/TreeMapper/components/TreeMapperList.tsx:53]: Will combine it

File: src/features/user/TreeMapper/components/Map.tsx (hunk 4 above)
Comment (sarveshpro) [reply src/features/user/TreeMapper/components/Map.tsx:132]: I have added try-catch, the values are NaN sometimes.

File: src/features/common/Layout/Footer/AccountFooter.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        if (countryCode) setSelectedCountry(countryCode);
+      }
+      if (localStorage.getItem('language')) {
+        const langCode = localStorage.getItem('language');
+        if (langCode) setLanguage(langCode);
+      }
+    }
+  }, []);
+  return ready ? (
+    <footer>
```
[the remark this answers is not in the document]
Comment (sarveshpro) [reply src/features/common/Layout/Footer/AccountFooter.tsx:50]: @sagararyal asked to remove the footer, language switcher is present in the sidebar, everything else is not needed.

**Comment by sarveshpro** [comment]
> * [ ]  My trees registration using a polygon is not displayed on the map - I assume we do not support this yet?
> 
It should can you share your profile slug and which environment was it? I'll check your public profile

**Comment by norbertschuler** [comment]
> > * [ ]   My trees registration using a polygon is not displayed on the map - I assume we do not support this yet?
> 
> It should can you share your profile slug and which environment was it? I'll check your public profile

- [ ] My profile link is not visible any more as the browser URL, so you tried sharing and found another bug that the URL in the sharing message is e.g. https://planet-webapp-4n8d4uop8-planetapp.vercel.app/profile without the users slug now.

**Comment by norbertschuler** [comment]
> > It should can you share your profile slug and which environment was it? I'll check your public profile

My profile not showing the blue tree registration: https://planet-webapp-4n8d4uop8-planetapp.vercel.app/t/norbert-schuler

File: src/features/user/TreeMapper/components/Map.tsx (hunk 4 above)
Comment (norbertschuler) [reply src/features/user/TreeMapper/components/Map.tsx:132]: @sarveshpro I could not see any commit from you regarding this change?

**Review by harshvitra — APPROVED** [review]
